<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Complete Parameter Reference &mdash; Yet Another Viral Subspecies Analysis Pipeline  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Output" href="output.html" />
    <link rel="prev" title="Usage" href="usage.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Yet Another Viral Subspecies Analysis Pipeline
            <img src="_static/yavsap_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Complete Parameter Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-output-options">Input/output options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reference-genome-options">Reference genome options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#kraken2-options">Kraken2 Options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#read-trimming-options">Read trimming options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#de-novo-assembly-options">de novo assembly options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#variant-calling-options">Variant calling options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#haplotype-calling-options">Haplotype calling options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#phylogenetic-tree-options">Phylogenetic tree options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#workflow-options">Workflow options</a></li>
<li class="toctree-l2"><a class="reference internal" href="#max-job-request-options">Max job request options</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="reference.html">Private API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Yet Another Viral Subspecies Analysis Pipeline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content style-external-links">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Complete Parameter Reference</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com//ksumngs/yavsap/blob/parameters.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="complete-parameter-reference">
<h1>Complete Parameter Reference<a class="headerlink" href="#complete-parameter-reference" title="Permalink to this headline"></a></h1>
<p>Every single parameter in the entire pipeline is described here. If you need to tweak every little detail of how your samples are analyzed, you’ve come to the right place! Note that <em>every</em> parameter is described here, even those that shouldn’t be used, so proceed with caution!</p>
<section id="input-output-options">
<h2>Input/output options<a class="headerlink" href="#input-output-options" title="Permalink to this headline"></a></h2>
<p>Define where the pipeline should find input data and save output data.</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">input</span></code></p></td>
<td><p>Path the folder containing input reads. <details><summary>Help</summary><small>For Illumina (paired-end) reads, the file names must be identical until the ending underscore with a read number, e.g. ‘sample1_S10_L001_R1_001.fastq.gz’ and ‘sample1_S10_L001_R2_001.fastq.gz’. The read number must be designated using either ‘_1’ and ‘_2’ or ‘_R1’ and ‘_R2’. For Nanopore reads, each fastq file is assumed to be a unique sample, so, e.g. ‘FAP01234_pass_barcode05_abcd01234_0.fastq.gz’ and ‘FAP01234_pass_barcode05_abcd01234_1.fastq.gz’ are assumed to be different samples even though they are from the same barcode. All read files must be gzipped, and have the extension ‘.fastq.gz’ or ‘.fq.gz’.</small></details></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">string</span></code></p></td>
<td><p>.</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">sra</span></code></p></td>
<td><p>Pull input reads from NCBI SRA <details><summary>Help</summary><small>This flag switches the meaning of <code class="docutils literal notranslate"><span class="pre">--input</span></code> to mean an NCBI Short Read Archive (SRA) accession number, then pulls the associated files directly from NCBI and runs the pipeline on them. To use this flag, you <strong>must</strong> have an <a class="reference external" href="https://ncbiinsights.ncbi.nlm.nih.gov/2017/11/02/new-api-keys-for-the-e-utilities/">NCBI API key</a> and it <strong>must</strong> be exported in the shell environment as <code class="docutils literal notranslate"><span class="pre">NCBI_API_KEY</span></code>, e.g. <code class="docutils literal notranslate"><span class="pre">NCBI_API_KEY=0123456789abcdef</span></code>. This feature is currenly broken due to an upstream issue in Nextflow, but should be fixed by Nextflow version 21.10.0.</small></details></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">platform</span></code></p></td>
<td><p>NGS platform used to sequence the samples</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">string</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">pe</span></code></p></td>
<td><p>INTERNAL USE ONLY!!!</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ont</span></code></p></td>
<td><p>INTERNAL USE ONLY!!!</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">seed</span></code></p></td>
<td><p>Set the random seed for all applicable processes</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>42</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">outdir</span></code></p></td>
<td><p>Path to the output directory where the results will be saved.</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">string</span></code></p></td>
<td><p>./results</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">publish_dir_mode</span></code></p></td>
<td><p>How to create results files</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">string</span></code></p></td>
<td><p>copy</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">tracedir</span></code></p></td>
<td><p>Directory to keep pipeline Nextflow logs and reports.</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">string</span></code></p></td>
<td><p>${params.outdir}/pipeline_info</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">help</span></code></p></td>
<td><p>Display help text.</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="reference-genome-options">
<h2>Reference genome options<a class="headerlink" href="#reference-genome-options" title="Permalink to this headline"></a></h2>
<p>Reference genome related files and options required for the workflow.</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">genome</span></code></p></td>
<td><p>NCBI accession number of reference genome <details><summary>Help</summary><small>If using a reference genome configured in the pipeline using iGenomes, use this parameter to give the ID for the reference. This is then used to build the full paths for all required reference genome files e.g. <code class="docutils literal notranslate"><span class="pre">--genome</span> <span class="pre">GRCh38</span></code>.</p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p>See the <a class="reference external" href="https://nf-co.re/usage/reference_genomes">nf-core website docs</a> for more details.</small></details>| <code class="docutils literal notranslate"><span class="pre">string</span></code> | NC_001437.1 |  |
| <code class="docutils literal notranslate"><span class="pre">genome_list</span></code> | Path to TSV file containing NCBI accession numbers of strains for constructing phylogenetic trees. | <code class="docutils literal notranslate"><span class="pre">string</span></code> | jev |  |</p>
</section>
<section id="kraken2-options">
<h2>Kraken2 Options<a class="headerlink" href="#kraken2-options" title="Permalink to this headline"></a></h2>
<p>Control how Kraken2 filters out host reads</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">kraken2_db</span></code></p></td>
<td><p>Path to a Kraken2 database <details><summary>Help</summary><small>The path to a <a class="reference external" href="https://github.com/DerrickWood/kraken2/wiki/Manual#custom-databases">Kraken2 database</a> that will be used to filter out host reads in the pipeline. This path will be automatically mounted into the container environments if a containerized profile is used.</p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p>Corresponds to the <a class="reference external" href="https://github.com/DerrickWood/kraken2/wiki/Manual#classification">–db</a> option of Kraken2.</small></details>| <code class="docutils literal notranslate"><span class="pre">string</span></code> | None |  |
| <code class="docutils literal notranslate"><span class="pre">keep_taxid</span></code> | Taxonomic IDs to keep and analyze <details><summary>Help</summary><small>A space-separated list (use quotes on the command line), of the taxonomic ids to keep based on Kraken2’s classification.</p>
<p>Defaults to keeping all unclassified reads and all viral reads. Note that this requires the host to be present in the Kraken2 database. When dealing with animals and the databases available from <code class="docutils literal notranslate"><span class="pre">kraken2-build</span></code>, this is not the case, and this parameter should be modified.</small></details>| <code class="docutils literal notranslate"><span class="pre">string</span></code> | 0 10239 |  |</p>
</section>
<section id="read-trimming-options">
<h2>Read trimming options<a class="headerlink" href="#read-trimming-options" title="Permalink to this headline"></a></h2>
<p>Options for how strictly to quality control NGS reads</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">trim_minlen</span></code></p></td>
<td><p>Minimum length of reads <details><summary>Help</summary><small>Corresponds to the <a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">MINLEN</a></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p>option of Trimmomatic for Illumina reads.</p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p>Corresponds to the <a class="reference external" href="https://github.com/wdecoster/nanofilt/#usage">–length</a>
option of NanoFilt for Nanopore reads.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 100 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_maxlen</span></code> | Maximum length of reads <details><summary>Help</summary><small>Only applies to Nanopore reads.</p>
<p>Corresponds to the <a class="reference external" href="https://github.com/wdecoster/nanofilt/#usage">–maxlength</a> option of
NanoFilt.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 0 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_adapters</span></code> | Sequences to be removed during trimming <details><summary>Help</summary><small>Only applies to Illumina reads. Corresponds to the first
<a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">ILLUMINACLIP</a> option of
Trimmomatic. If left blank (i.e. <code class="docutils literal notranslate"><span class="pre">--trim_adapters</span> <span class="pre">''</span></code>), then adapter trimming is
disabled. Custom adapters cannot be used, and the parameter corresponds to
one of the prebuilt sequence files provided with Trimmomatic.
</small></details>| <code class="docutils literal notranslate"><span class="pre">string</span></code> | NexteraPE-PE.fa |  |
| <code class="docutils literal notranslate"><span class="pre">trim_mismatches</span></code> | Max number of base mismatches to allow an adapter match <details><summary>Help</summary><small>Only applies to Illumina reads. Corresponds to the second
<a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">ILLUMINACLIP</a> option of
Trimmomatic.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 2 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_pclip</span></code> | How accurate the match between adapter ligated reads must be for paired-end palindrome read alignment <details><summary>Help</summary><small>Only applies to Illumina reads. Corresponds to the third
<a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">ILLUMINACLIP</a> option of
Trimmomatic.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 30 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_clip</span></code> | How accurate the match between any adapter must be against a read <details><summary>Help</summary><small>Only applies to Illumina reads. Corresponds to the final
<a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">ILLUMINACLIP</a> option of
Trimmomatic.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 10 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_winsize</span></code> | Number of bases to average quality across <details><summary>Help</summary><small>Only applies to Illumina reads. If set to <code class="docutils literal notranslate"><span class="pre">0</span></code>, then sliding window trimming is disabled.
Corresponds to the first
<a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">SLIDINGWINDOW</a> option of
Trimmomatic.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 50 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_winqual</span></code> | Required average window base quality <details><summary>Help</summary><small>Only applies to Illumina reads. If set to <code class="docutils literal notranslate"><span class="pre">0</span></code>, then sliding window trimming is disabled.
Corresponds to the second
<a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">SLIDINGWINDOW</a> option of
Trimmomatic.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 15 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_leading</span></code> | Minimum quality of bases in leading end of read <details><summary>Help</summary><small>Only applies to Illumina reads. If set to <code class="docutils literal notranslate"><span class="pre">0</span></code>, LEADING trimming is disabled.
Corresponds to the
<a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">LEADING</a> option of
Trimmomatic.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 15 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_trailing</span></code> | Minimum quality of bases in trailing end of read <details><summary>Help</summary><small>Only applies to Illumina reads. If set to <code class="docutils literal notranslate"><span class="pre">0</span></code>, TRAILING trimming is disabled.
Corresponds to the <a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">TRAILING</a> option of
Trimmomatic.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 15 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_headcrop</span></code> | Number of bases to remove from start of read <details><summary>Help</summary><small>Corresponds to the <a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">HEADCROP</a>
option of Trimmomatic for Illumina reads. If set to <code class="docutils literal notranslate"><span class="pre">0</span></code>, then HEADCROP trimming is disabled.</p>
<p>Corresponds to the <a class="reference external" href="https://github.com/wdecoster/nanofilt/#usage">–headcrop</a>
option of NanoFilt for Nanopore reads.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 0 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_crop</span></code> | Number of bases to keep from start of read <details><summary>Help</summary><small>Only applies to Illumina reads. If set to <code class="docutils literal notranslate"><span class="pre">0</span></code>, CROP trimming is disabled.
Corresponds to the
<a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">CROP</a> option of Trimmomatic.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 0 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_meanqual</span></code> | Minimum average base quality of entire reads <details><summary>Help</summary><small>Applies only to ONT reads. Corresponds to the
<a class="reference external" href="https://github.com/wdecoster/nanofilt/#usage">–quality</a> option of NanoFilt.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 7 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_mingc</span></code> | Minimum GC count of reads <details><summary>Help</summary><small>Only applies to ONT reads. Corresponds to the
<a class="reference external" href="https://github.com/wdecoster/nanofilt/#usage">–minGC</a> option of NanoFilt.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 0 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_maxgc</span></code> | Maximum GC count of reads <details><summary>Help</summary><small>Only applies to ONT reads. Corresponds to the
<a class="reference external" href="https://github.com/wdecoster/nanofilt/#usage">–maxGC</a> option of NanoFilt.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 0 |  |
| <code class="docutils literal notranslate"><span class="pre">trim_tailcrop</span></code> | Number of bases to remove from the end of each read <details><summary>Help</summary><small>Only applies to ONT reads. Corresponds to the
<a class="reference external" href="https://github.com/wdecoster/nanofilt/#usage">–tailcrop</a> option of NanoFilt.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 0 |  |</p>
</section>
<section id="de-novo-assembly-options">
<h2>de novo assembly options<a class="headerlink" href="#de-novo-assembly-options" title="Permalink to this headline"></a></h2>
<p>Options for how to construct contigs</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">spades_mode</span></code></p></td>
<td><p>Type of reads to assemble using SPAdes <details><summary>Help</summary><small>Only applies to Illumina reads. This parameter is turned into a flag and passed as the ‘mode’ to</p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p>the SPAdes assembly. Note that the <code class="docutils literal notranslate"><span class="pre">isolate</span></code> and <code class="docutils literal notranslate"><span class="pre">bio</span></code> modes normally provided by SPAdes</p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>are unavailable in the pipeline.</p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p>See <a class="reference external" href="https://cab.spbu.ru/files/release3.15.3/manual.html#sec3.2">SPAdes command line options</a>
for more info on what each of these mean.</small></details>| <code class="docutils literal notranslate"><span class="pre">string</span></code> | rnaviral |  |
| <code class="docutils literal notranslate"><span class="pre">canu_corrected_error_rate</span></code> | How dissimilar two reads can be and still be assembled together <details><summary>Help</summary><small>Only applies to ONT reads. Corresponds to the <span class="xref std std-ref">correctedErrorRate</span> option of Canu.</small></details>| <code class="docutils literal notranslate"><span class="pre">number</span></code> | 0.144 |  |
| <code class="docutils literal notranslate"><span class="pre">canu_min_read_length</span></code> | Minimum length allowed for assembled reads <details><summary>Help</summary><small>Only applies to ONT reads.
Corresponds to the <span class="xref std std-ref">minReadLength</span> option of Canu.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 1000 |  |
| <code class="docutils literal notranslate"><span class="pre">canu_min_overlap_length</span></code> | Shortest overlap for reads to be assembled together <details><summary>Help</summary><small>Only applies to ONT reads.
Corresponds to the <span class="xref std std-ref">minOverlapLength</span> option of Canu.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 500 |  |
| <code class="docutils literal notranslate"><span class="pre">canu_stop_on_low_coverage</span></code> | Minimum depth to allow assembly <details><summary>Help</summary><small>Only applies to ONT reads.
Corresponds to the <span class="xref std std-ref">stopOnLowCoverage</span> option of Canu.</small></details>| <code class="docutils literal notranslate"><span class="pre">integer</span></code> | 10 |  |</p>
</section>
<section id="variant-calling-options">
<h2>Variant calling options<a class="headerlink" href="#variant-calling-options" title="Permalink to this headline"></a></h2>
<p>Options for how strict to call variants</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">variant_quality</span></code></p></td>
<td><p>Minimum average quality (PHRED score) of a basecall to consider it a variant</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>12</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">variant_depth</span></code></p></td>
<td><p>Minimum depth to consider a non-reference basecall a variant</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>15</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">variant_position</span></code></p></td>
<td><p>Do not consider variants that occur only in positions within this percentage of the end</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">number</span></code></p></td>
<td><p>0.1</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">variant_frequency</span></code></p></td>
<td><p>Minimum frequency at which a variant must appear</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">number</span></code></p></td>
<td><p>0.05</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">variant_significance</span></code></p></td>
<td><p>Confidence threshold for variants based on Fisher’s Exact test</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">number</span></code></p></td>
<td><p>0.001</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="haplotype-calling-options">
<h2>Haplotype calling options<a class="headerlink" href="#haplotype-calling-options" title="Permalink to this headline"></a></h2>
<p>Options for how strictly to call haplotypes</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">haplotype_depth</span></code></p></td>
<td><p>Minmum number of times a haplotype must appear</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>10</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">haplotype_frequency</span></code></p></td>
<td><p>Minimum haplotype frequency</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">number</span></code></p></td>
<td><p>0.05</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">haplotype_significance</span></code></p></td>
<td><p>Confidence threshold for haplotypes based on a X-squared test of linkage disequilibrium</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">number</span></code></p></td>
<td><p>0.05</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">haplotype_method</span></code></p></td>
<td><p>How to construct pseudo-reads for haplotyping</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">string</span></code></p></td>
<td><p>ml-template</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">haplotype_overlap_min</span></code></p></td>
<td><p>Minimum number of bases reads must overlap to be built into the same pseudo-read</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>0</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">haplotype_overlap_max</span></code></p></td>
<td><p>Maximum number of bases reads must overlap to be built into the same pseudo-read</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>100</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">haplotype_iterations</span></code></p></td>
<td><p>Number of iterations to build pseudo-reads</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>1000</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="phylogenetic-tree-options">
<h2>Phylogenetic tree options<a class="headerlink" href="#phylogenetic-tree-options" title="Permalink to this headline"></a></h2>
<p>Options to control how the phylogenetic trees are generated</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">phylogenetic_bootstraps</span></code></p></td>
<td><p>Number of iterations to create phylogenetic trees with</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>1000</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">phylogenetic_bootstrap_cutoff</span></code></p></td>
<td><p>Minimum quality for phylogenetic bootstraps</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">number</span></code></p></td>
<td><p>0.03</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="workflow-options">
<h2>Workflow options<a class="headerlink" href="#workflow-options" title="Permalink to this headline"></a></h2>
<p>Options to skip portions of the workflow</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">skip_trimming</span></code></p></td>
<td><p>Skip read trimming (Trimmomatic/Nanofilt)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">skip_qc</span></code></p></td>
<td><p>Skip FastQC</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">skip_filtering</span></code></p></td>
<td><p>Skip Kraken2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">skip_assembly</span></code></p></td>
<td><p>Skip <em>de novo</em> assembly (SPAdes/canu)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">skip_haplotype</span></code></p></td>
<td><p>Skip haplotyping (CliqueSNV/HapLink)</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">boolean</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="max-job-request-options">
<h2>Max job request options<a class="headerlink" href="#max-job-request-options" title="Permalink to this headline"></a></h2>
<p>Set the top limit for requested resources for any single job.</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Parameter</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>Type</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Required</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">max_cpus</span></code></p></td>
<td><p>Maximum number of CPUs that can be requested for any single job. <details><summary>Help</summary><small>Use to set an upper-limit for the CPU requirement for each process. Should be an integer e.g. <code class="docutils literal notranslate"><span class="pre">--max_cpus</span> <span class="pre">1</span></code></small></details></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">integer</span></code></p></td>
<td><p>16</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">max_memory</span></code></p></td>
<td><p>Maximum amount of memory that can be requested for any single job. <details><summary>Help</summary><small>Use to set an upper-limit for the memory requirement for each process. Should be a string in the format integer-unit e.g. <code class="docutils literal notranslate"><span class="pre">--max_memory</span> <span class="pre">'8.GB'</span></code></small></details></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">string</span></code></p></td>
<td><p>128.GB</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">max_time</span></code></p></td>
<td><p>Maximum amount of time that can be requested for any single job. <details><summary>Help</summary><small>Use to set an upper-limit for the time requirement for each process. Should be a string in the format integer-unit e.g. <code class="docutils literal notranslate"><span class="pre">--max_time</span> <span class="pre">'2.h'</span></code></small></details></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">string</span></code></p></td>
<td><p>240.h</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="usage.html" class="btn btn-neutral float-left" title="Usage" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="output.html" class="btn btn-neutral float-right" title="Output" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Thomas A. Christensen II.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  
<div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    <span class="fa fa-book"> Other Versions</span>
    v: v0.4.0-alpha
    <span class="fa fa-caret-down"></span>
  </span>
  <div class="rst-other-versions">
    <dl>
      <dt>Tags</dt>
      <dd><a href="../v0.1.0-alpha/parameters.html">v0.1.0-alpha</a></dd>
      <dd><a href="../v0.1.1/parameters.html">v0.1.1</a></dd>
      <dd><a href="../v0.2.0-alpha/parameters.html">v0.2.0-alpha</a></dd>
      <dd><a href="../v0.2.1-alpha/parameters.html">v0.2.1-alpha</a></dd>
      <dd><a href="../v0.3.0-alpha/parameters.html">v0.3.0-alpha</a></dd>
      <dd><a href="../v0.3.1-alpha/parameters.html">v0.3.1-alpha</a></dd>
      <dd><a href="../v0.3.2-alpha/parameters.html">v0.3.2-alpha</a></dd>
      <dd><a href="../v0.3.3-alpha/parameters.html">v0.3.3-alpha</a></dd>
      <dd><a href="parameters.html">v0.4.0-alpha</a></dd>
      <dd><a href="../v0.5.0-alpha/index.html">v0.5.0-alpha</a></dd>
      <dd><a href="../v0.6.0-alpha/index.html">v0.6.0-alpha</a></dd>
      <dd><a href="../v0.6.1-alpha/index.html">v0.6.1-alpha</a></dd>
      <dd><a href="../v0.6.2-alpha/index.html">v0.6.2-alpha</a></dd>
      <dd><a href="../v0.6.3-alpha/index.html">v0.6.3-alpha</a></dd>
      <dd><a href="../v0.6.4-alpha/index.html">v0.6.4-alpha</a></dd>
      <dd><a href="../v0.6.5-alpha/index.html">v0.6.5-alpha</a></dd>
      <dd><a href="../v0.7.0/index.html">v0.7.0</a></dd>
      <dd><a href="../v0.7.1/index.html">v0.7.1</a></dd>
      <dd><a href="../v0.8.0/index.html">v0.8.0</a></dd>
    </dl>
    <dl>
      <dt>Branches</dt>
      <dd><a href="../develop/index.html">develop</a></dd>
    </dl>
  </div>
</div><script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>