<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Usage &mdash; Yet Another Viral Subspecies Analysis Pipeline  documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Complete Parameter Reference" href="parameters.html" />
    <link rel="prev" title="Installation" href="install.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Yet Another Viral Subspecies Analysis Pipeline
            <img src="_static/yavsap_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Usage</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#input-preparation">Input preparation</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#using-a-directory-as-input">Using a Directory as Input</a></li>
<li class="toctree-l3"><a class="reference internal" href="#using-a-samplesheet-as-input">Using a Samplesheet as Input</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#single-end-example">Single-end example</a></li>
<li class="toctree-l4"><a class="reference internal" href="#paired-end-example">Paired-end example</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#running-the-pipeline">Running the pipeline</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#updating-the-pipeline">Updating the pipeline</a></li>
<li class="toctree-l3"><a class="reference internal" href="#reproducibility">Reproducibility</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#core-nextflow-arguments">Core Nextflow arguments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#profile"><code class="docutils literal notranslate"><span class="pre">-profile</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#resume"><code class="docutils literal notranslate"><span class="pre">-resume</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#c"><code class="docutils literal notranslate"><span class="pre">-c</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#mandatory-pipeline-parameters">Mandatory pipeline parameters</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genome-preparation">Genome Preparation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#running-in-the-background">Running in the background</a></li>
<li class="toctree-l2"><a class="reference internal" href="#nextflow-memory-requirements">Nextflow memory requirements</a></li>
<li class="toctree-l2"><a class="reference internal" href="#setting-up-for-hpc-job-schedulers">Setting up for HPC job schedulers</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#process-labels">Process Labels</a></li>
<li class="toctree-l3"><a class="reference internal" href="#example">Example</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="parameters.html">Complete Parameter Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">Output</a></li>
<li class="toctree-l1"><a class="reference internal" href="report.html">The YAVSAP report</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Yet Another Viral Subspecies Analysis Pipeline</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content style-external-links">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Usage</li>
      <li class="wy-breadcrumbs-aside">
              <a href="https://github.com//ksumngs/yavsap/blob/usage.md" class="fa fa-github"> Edit on GitHub</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <!--markdownlint-disable line-length -->
<section id="usage">
<h1>Usage<a class="headerlink" href="#usage" title="Permalink to this headline"></a></h1>
<p>Basic jist:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow run ksumngs/yavsap <span class="se">\</span>
  -profile &lt;docker,podman,singularity,INSTITUTE&gt; <span class="se">\</span>
  --platform &lt;illumina,nanopore&gt; --kraken2_db &lt;kraken2_db&gt;
</pre></div>
</div>
<section id="input-preparation">
<h2>Input preparation<a class="headerlink" href="#input-preparation" title="Permalink to this headline"></a></h2>
<section id="using-a-directory-as-input">
<h3>Using a Directory as Input<a class="headerlink" href="#using-a-directory-as-input" title="Permalink to this headline"></a></h3>
<p>If the <a class="reference internal" href="parameters.html#input-output-options"><span class="std std-ref">–input</span></a> parameter is not passed, then
YAVSAP assumes reads files are located in the current directory. You can also
pass a directory that contains reads files to <code class="docutils literal notranslate"><span class="pre">--input</span></code>. YAVSAP will not look
in subdirectories of <code class="docutils literal notranslate"><span class="pre">--input</span></code>. When running YAVSAP on a directory, Reads
files must have one of one of the following extensions:</p>
<ul class="simple">
<li><p>.fastq</p></li>
<li><p>.fq</p></li>
<li><p>.fastq.gz</p></li>
<li><p>.fq.gz</p></li>
</ul>
<p>There must be only one file per sample for single-end and inteleaved paired-end
reads, and one pair of files (two files) for regular paired-end reads. Regular
paired-end files must have either the characters <code class="docutils literal notranslate"><span class="pre">_1</span></code> and <code class="docutils literal notranslate"><span class="pre">_2</span></code> or <code class="docutils literal notranslate"><span class="pre">_R1</span></code>
and <code class="docutils literal notranslate"><span class="pre">_R2</span></code> in the filename to be identified as a pair. The sample name is taken
as the all the characters <em>before</em> the first underscore in the file name.</p>
<p>For example, a folder with the following files</p>
<div class="highlight-text notranslate"><div class="highlight"><pre><span></span>.
├── pig-serum_S4_L001_R1_001.fastq.gz
├── pig-serum_S4_L001_R2_001.fastq.gz
├── pig-feces_S5_L001_R1_001.fastq.gz
├── pig-feces_S5_L001_R2_001.fastq.gz
├── mosquito1_S6_L001_R1_001.fastq.gz
├── mosquito1_S6_L001_R2_001.fastq.gz
├── mosquito2_S7_L001_R1_001.fastq.gz
└── mosquito2_S7_L001_R2_001.fastq.gz
</pre></div>
</div>
<p>will produce a sample list of</p>
<ul class="simple">
<li><p>pig-serum</p></li>
<li><p>pig-feces</p></li>
<li><p>mosquito1</p></li>
<li><p>mosquito2</p></li>
</ul>
<p>Paired-endedness is determined based on the <a class="reference internal" href="parameters.html#input-output-options"><span class="std std-ref">–paired</span></a> flag. By default <code class="docutils literal notranslate"><span class="pre">--paired</span></code> is <code class="docutils literal notranslate"><span class="pre">true</span></code> when <a class="reference internal" href="parameters.html#input-output-options"><span class="std std-ref">–platform</span></a> is <code class="docutils literal notranslate"><span class="pre">illumina</span></code> and <code class="docutils literal notranslate"><span class="pre">false</span></code> for <code class="docutils literal notranslate"><span class="pre">nanopore</span></code>, but this
can be overridden. To use interleaved paired-end reads, use the
<a class="reference internal" href="parameters.html#input-output-options"><span class="std std-ref">–interleaved</span></a> flag.</p>
<p>This file naming system is very common for Illumina output, however for Oxford
Nanopore reads, the default file structure breaks this in several ways, and it is
often better to use a samplesheet.</p>
</section>
<section id="using-a-samplesheet-as-input">
<h3>Using a Samplesheet as Input<a class="headerlink" href="#using-a-samplesheet-as-input" title="Permalink to this headline"></a></h3>
<p>A samplesheet allows you to pass multiple reads files in as
a single sample, which can be useful for e.g. Oxford Nanopore reads. When using
a samplesheet, simply pass the path to the sheet to <a class="reference internal" href="parameters.html#input-output-options"><span class="std std-ref">–input</span></a>.</p>
<p>Samplesheets are tab-delimited files. A header is optional, and must be
delineated by beginning with the pound (<code class="docutils literal notranslate"><span class="pre">#</span></code>) symbol. The first column contains
sample names, and each subsequent column contains a path to reads files. All
sample names will be cleaned of any shell metacharacters, and will be truncated
at the first underscore. Paths are resolved relative to the current directory.
Path names can contain wildcard characters, and will resolve as glob patterns
identical to how they would in Nextflow’s <a class="reference external" href="https://nextflow.io/docs/latest/channel.html#channel-path" title="(in Nextflow v22.04)"><span class="xref std std-ref">Channel.fromPath</span></a>
constructor. For paired-end reads, forward reads should be specified in the
even-number columns (<code class="docutils literal notranslate"><span class="pre">2,4,6</span></code>), and reverse reads should be specified in the
odd-number columns (<code class="docutils literal notranslate"><span class="pre">3,5,7</span></code>).</p>
<section id="single-end-example">
<h4>Single-end example<a class="headerlink" href="#single-end-example" title="Permalink to this headline"></a></h4>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>#samplename</p></th>
<th class="head"><p></p></th>
<th class="head"><p></p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>pig-serum</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/data/run/fastq{pass,fail}/barcode01/FAP01234_{pass,fail}_barcode01_abcde01_*.fastq*</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/data/run2/fastq{pass,fail}/barcode07/FAP01234_{pass,fail}_barcode07_abcde01_*.fastq*</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>pig-feces</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/data/run/fastq{pass,fail}/barcode02/FAP01234_{pass,fail}_barcode02_abcde01_*.fastq*</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>mosquito1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/data/run/fastq{pass,fail}/barcode03/FAP01234_{pass,fail}_barcode03_abcde01_*.fastq*</span></code></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p>mosquito2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">./seq-results/mosquito2/*.fastq*`</span></code></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="paired-end-example">
<h4>Paired-end example<a class="headerlink" href="#paired-end-example" title="Permalink to this headline"></a></h4>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>#Sample</p></th>
<th class="head"><p>Forward1</p></th>
<th class="head"><p>Reverse1</p></th>
<th class="head"><p>Forward2</p></th>
<th class="head"><p>Reverse2</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>pig-serum</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/basespace/run/PIG-SERUM*_R1*.fastq.gz</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/basespace/run/PIG-SERUM*_R2*.fastq.gz</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/dragen/run/PIG-SERUM*_R1*.fastq.gz</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/dragen/run/PIG-SERUM*_R2*.fastq.gz</span></code></p></td>
</tr>
<tr class="row-odd"><td><p>pig-feces</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/basespace/run/PIG-FECES*_R1*.fastq.gz</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/basespace/run/PIG-FECES*_R2*.fastq.gz</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p>mosquito1</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/basespace/run/MOSQUITO1*_R1*.fastq.gz</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">/basespace/run/MOSQUITO1*_R1*.fastq.gz</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p>mosquito2</p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">./seq-results/mosquito2/*_R1*.fastq.gz</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">./seq-results/mosquito2/*_R2*.fastq.gz</span></code></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p>Once the samplesheet is constructed, pass it on the command line as:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>nextflow run ksumngs/yavsap <span class="se">\</span>
  -profile &lt;docker,podman,singularity,INSTITUTE&gt; <span class="se">\</span>
  --input /path/to/sheet.tsv <span class="se">\</span>
  --platform &lt;illumina,nanopore&gt; <span class="se">\</span>
  --kraken2_db &lt;kraken2_db&gt;
</pre></div>
</div>
</section>
</section>
</section>
<section id="running-the-pipeline">
<h2>Running the pipeline<a class="headerlink" href="#running-the-pipeline" title="Permalink to this headline"></a></h2>
<p>The typical command for running the pipeline is as follows:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow run ksumngs/yavsap -profile docker --input . --outdir &lt;OUTDIR&gt; --platform illumina --kraken2_db https://genome-idx.s3.amazonaws.com/kraken/k2_viral_20210517.tar.gz --keep_taxid classified</span>
</pre></div>
</div>
<p>This will launch the pipeline with the <code class="docutils literal notranslate"><span class="pre">docker</span></code> configuration profile. See below for more information about profiles.</p>
<p>Note that the pipeline will create the following files in your working directory:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">work                # Directory containing the nextflow working files</span>
<span class="go">&lt;OUTIDR&gt;            # Finished results in specified location (defined with --outdir)</span>
<span class="go">.nextflow_log       # Log file from Nextflow</span>
<span class="gp"># </span>Other nextflow hidden files, eg. <span class="nb">history</span> of pipeline runs and old logs.
</pre></div>
</div>
<section id="updating-the-pipeline">
<h3>Updating the pipeline<a class="headerlink" href="#updating-the-pipeline" title="Permalink to this headline"></a></h3>
<p>When you run the above command, Nextflow automatically pulls YAVSAP from GitHub and stores it as a cached version. When running the pipeline after this, it will always use the cached version if available - even if we’ve updated it since. To make sure that you’re running the latest version of the pipeline, you can update the cached version of the pipeline:</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">nextflow pull ksumngs/yavsap</span>
</pre></div>
</div>
</section>
<section id="reproducibility">
<h3>Reproducibility<a class="headerlink" href="#reproducibility" title="Permalink to this headline"></a></h3>
<p>It is often a good idea to specify a pipeline version when running the pipeline on your data. This ensures that a specific version of the pipeline code and software are used when you run your pipeline. If you keep using the same tag, you’ll be running the same version of the pipeline, even if there have been changes to the code since.</p>
<p>First, go to the <a class="reference external" href="https://github.com/ksumngs/yavsap/releases">yavsap releases page</a> and find the latest version number (eg. <code class="docutils literal notranslate"><span class="pre">v1.3.1</span></code>). Then specify this when running the pipeline with <code class="docutils literal notranslate"><span class="pre">-r</span></code> (one hyphen) - eg. <code class="docutils literal notranslate"><span class="pre">-r</span> <span class="pre">v1.3.1</span></code>.</p>
<p>This version number will be logged in <a class="reference internal" href="report.html#the-yavsap-report"><span class="std std-ref">The YAVSAP report</span></a> when you run the pipeline, so that you’ll know what you used when you look back in the future.</p>
</section>
</section>
<section id="core-nextflow-arguments">
<h2>Core Nextflow arguments<a class="headerlink" href="#core-nextflow-arguments" title="Permalink to this headline"></a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>These options are part of Nextflow and use a <em>single</em> hyphen (pipeline parameters use a double-hyphen).</p>
</div>
<section id="profile">
<h3><code class="docutils literal notranslate"><span class="pre">-profile</span></code><a class="headerlink" href="#profile" title="Permalink to this headline"></a></h3>
<p>Use this parameter to choose a configuration profile. Profiles can give configuration presets for different compute environments.</p>
<p>Several generic profiles are bundled with the pipeline which instruct the pipeline to use software packaged using different methods (Docker, Singularity, Podman, Shifter, Charliecloud, Conda) - see below. When using Biocontainers, most of these software packaging methods pull Docker containers from <a class="reference external" href="https://quay.io">quay.io</a>. Singularity instead directly downloads Singularity images via https hosted by the <a class="reference external" href="https://depot.galaxyproject.org/singularity/">Galaxy project</a>, though.<!-- and Conda which downloads and installs software locally from [Bioconda](https://bioconda.github.io/)-->.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>We highly recommend the use of Docker or Singularity containers for full pipeline reproducibility: they just work!<!--, however when this is not possible, Conda is also supported-->.</p>
</div>
<p>The pipeline also dynamically loads configurations from <a class="reference external" href="https://github.com/nf-core/configs">https://github.com/nf-core/configs</a> when it runs, making multiple config profiles for various institutional clusters available at run time. For more information and to see if your system is available in these configs please see the <a class="reference external" href="https://github.com/nf-core/configs#documentation">nf-core/configs documentation</a>.</p>
<p>Note that multiple profiles can be loaded, for example: <code class="docutils literal notranslate"><span class="pre">-profile</span> <span class="pre">test,docker</span></code> - the order of arguments is important!
They are loaded in sequence, so later profiles can overwrite earlier profiles.</p>
<p>If <code class="docutils literal notranslate"><span class="pre">-profile</span></code> is not specified, the pipeline will run locally and expect all software to be installed and available on the <code class="docutils literal notranslate"><span class="pre">PATH</span></code>. This is <em>not</em> recommended.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">docker</span></code></p>
<ul>
<li><p>A generic configuration profile to be used with <a class="reference external" href="https://docker.com/">Docker</a></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">singularity</span></code></p>
<ul>
<li><p>A generic configuration profile to be used with <a class="reference external" href="https://sylabs.io/docs/">Singularity</a></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">podman</span></code></p>
<ul>
<li><p>A generic configuration profile to be used with <a class="reference external" href="https://podman.io/">Podman</a></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">shifter</span></code></p>
<ul>
<li><p>A generic configuration profile to be used with <a class="reference external" href="https://nersc.gitlab.io/development/shifter/how-to-use/">Shifter</a></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">charliecloud</span></code></p>
<ul>
<li><p>A generic configuration profile to be used with <a class="reference external" href="https://hpc.github.io/charliecloud/">Charliecloud</a></p></li>
</ul>
</li>
<li><p><s><code class="docutils literal notranslate"><span class="pre">conda</span></code></s></p>
<ul>
<li><p><s>A generic configuration profile to be used with <a class="reference external" href="https://conda.io/docs/">Conda</a>. Please only use Conda as a last resort i.e. when it’s not possible to run the pipeline with Docker, Singularity, Podman, Shifter or Charliecloud.</s> <strong>YAVSAP doesn’t currently have conda support, but we’re working on it!. Don’t use this profile!</strong></p></li>
</ul>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">test</span></code></p>
<ul>
<li><p>A profile with a complete configuration for automated testing</p></li>
<li><p>Includes links to test data so needs no other parameters</p></li>
</ul>
</li>
</ul>
</section>
<section id="resume">
<h3><code class="docutils literal notranslate"><span class="pre">-resume</span></code><a class="headerlink" href="#resume" title="Permalink to this headline"></a></h3>
<p>Specify this when restarting a pipeline. Nextflow will use cached results from any pipeline steps where the inputs are the same, continuing from where it got to previously. For input to be considered the same, not only the names must be identical but the files’ contents as well. For more info about this parameter, see <a class="reference external" href="https://www.nextflow.io/blog/2019/demystifying-nextflow-resume.html">this blog post</a>.</p>
<p>You can also supply a run name to resume a specific run: <code class="docutils literal notranslate"><span class="pre">-resume</span> <span class="pre">[run-name]</span></code>. Use the <code class="docutils literal notranslate"><span class="pre">nextflow</span> <span class="pre">log</span></code> command to show previous run names.</p>
</section>
<section id="c">
<h3><code class="docutils literal notranslate"><span class="pre">-c</span></code><a class="headerlink" href="#c" title="Permalink to this headline"></a></h3>
<p>Specify the path to a specific config file (this is a core Nextflow command). See the <a class="reference external" href="https://nf-co.re/usage/configuration">nf-core website documentation</a> for more information.</p>
</section>
</section>
<section id="mandatory-pipeline-parameters">
<h2>Mandatory pipeline parameters<a class="headerlink" href="#mandatory-pipeline-parameters" title="Permalink to this headline"></a></h2>
<p>See <a class="reference internal" href="parameters.html"><span class="doc">the page on parameters</span></a> for the complete lowdown on
parameters.</p>
<p>The pipeline is pretty much set up to run itself. So long as you have your input
reads formatted correctly, it doesn’t need much input from you. (Provided, the
defaults are for analyzing Japanese Encephalitis Virus, so your results might be
way off if you don’t have any JEV in your samples, see <a class="reference internal" href="#genome-preparation"><span class="std std-ref">Genome Preparation</span></a>.) These are the bare minimum parameters that you must provide on
the command-line for the pipeline to complete. Note that there has been mixed
success with placing these parameters in a <code class="docutils literal notranslate"><span class="pre">nextflow.config</span></code> file, so keeping
them on the command-line is best.</p>
<dl class="simple myst">
<dt>–kraken2_db</dt><dd><p>The path to a Kraken2 database. See <a class="reference internal" href="parameters.html#kraken2-options"><span class="std std-ref">–kraken2_db</span></a>.</p>
</dd>
<dt>–platform</dt><dd><p>Must be set to <code class="docutils literal notranslate"><span class="pre">illumina</span></code> or <code class="docutils literal notranslate"><span class="pre">nanopore</span></code>, depending on the type of reads
you are analyzing. See <a class="reference internal" href="parameters.html#input-output-options"><span class="std std-ref">–platform</span></a>.</p>
</dd>
</dl>
</section>
<section id="genome-preparation">
<h2>Genome Preparation<a class="headerlink" href="#genome-preparation" title="Permalink to this headline"></a></h2>
<p>YAVSAP needs some idea of what kind of viruses it’s looking for. You will need
to provide it with</p>
<ol class="arabic simple">
<li><p>A reference genome of the virus of interest</p></li>
<li><p>Genome examples of different strains or relatives of the virus of interest</p></li>
</ol>
<p>YAVSAP needs these in the form of NCBI GenBank Accession Numbers. To pick a new
reference genome, simply pass the accession number to the <a class="reference internal" href="parameters.html#reference-genome-options"><span class="std std-ref">–genome</span></a> parameter, e.g. <code class="docutils literal notranslate"><span class="pre">nextflow</span> <span class="pre">run</span> <span class="pre">ksumngs/yavsap</span> <span class="pre">--genome</span> <span class="pre">'NC_001437.1'</span></code>.</p>
<p>Close relative examples must be provided as a tab-delimited document with the
name in the first column, and the NCBI accession number in the second column. No
header is allowed in the comparison genomes file.</p>
<p>Nodes will be grouped in the output phylogenetic tree based on the name before
any underscore, to allow for easy visualization of known related groups. In
other words, <code class="docutils literal notranslate"><span class="pre">Nakayama\tEF571853.1</span></code> and <code class="docutils literal notranslate"><span class="pre">SA14\tMH258848.1</span></code> are perfectly
valid entries in the relatives list, but <code class="docutils literal notranslate"><span class="pre">GIII_Nakayama\tEF571853.1</span></code> and
<code class="docutils literal notranslate"><span class="pre">GIII_SA14\tMH258848.1</span></code> will give these two entries the same color so you can
easily see the Genotype III node.</p>
<p>There <strong>MUST</strong> be one entry whose name begins with <code class="docutils literal notranslate"><span class="pre">ROOT</span></code>. The genome of this
entry will be used as the outgroup for the resulting phylogenetic tree.</p>
<p>Once a comparison genomes file is prepared, the path to it can be passed to the
<a class="reference internal" href="parameters.html#reference-genome-options"><span class="std std-ref">–genome_list</span></a> parameter.</p>
<p>YAVSAP comes with some example comparison genomes files. These can be referred
to by name, rather than by path(e.g. <code class="docutils literal notranslate"><span class="pre">nextflow</span> <span class="pre">run</span> <span class="pre">ksumngs/yavsap</span> <span class="pre">--genome_list</span> <span class="pre">jev</span></code>). They are</p>
<table class="colwidths-auto docutils align-default">
<thead>
<tr class="row-odd"><th class="head"><p>Virus</p></th>
<th class="head"><p>Name</p></th>
<th class="head"><p>Recommended Reference</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/data-hub/taxonomy/11072/">Japanese Encephalitis Virus (JEV)</a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">jev</span></code></p></td>
<td><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/NC_001437.1">NC_001437.1</a></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/labs/data-hub/taxonomy/2697049/">Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)</a></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">sc2</span></code></p></td>
<td><p><a class="reference external" href="https://www.ncbi.nlm.nih.gov/nuccore/NC_045512.2">NC_045512.2</a></p></td>
</tr>
</tbody>
</table>
</section>
<section id="running-in-the-background">
<h2>Running in the background<a class="headerlink" href="#running-in-the-background" title="Permalink to this headline"></a></h2>
<p>Nextflow handles job submissions and supervises the running jobs. The Nextflow process must run until the pipeline is finished.</p>
<p>The Nextflow <code class="docutils literal notranslate"><span class="pre">-bg</span></code> flag launches Nextflow in the background, detached from your terminal so that the workflow does not stop if you log out of your session. The logs are saved to a file.</p>
<p>Alternatively, you can use <code class="docutils literal notranslate"><span class="pre">screen</span></code> / <code class="docutils literal notranslate"><span class="pre">tmux</span></code> or similar tool to create a detached session which you can log back into at a later time.
Some HPC setups also allow you to run nextflow within a cluster job submitted your job scheduler (from where it submits more jobs).</p>
</section>
<section id="nextflow-memory-requirements">
<h2>Nextflow memory requirements<a class="headerlink" href="#nextflow-memory-requirements" title="Permalink to this headline"></a></h2>
<p>In some cases, the Nextflow Java virtual machines can start to request a large amount of memory.
We recommend adding the following line to your environment to limit this (typically in <code class="docutils literal notranslate"><span class="pre">~/.bashrc</span></code> or <code class="docutils literal notranslate"><span class="pre">~./bash_profile</span></code>):</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">NXF_OPTS=&#39;-Xms1g -Xmx4g&#39;</span>
</pre></div>
</div>
</section>
<section id="setting-up-for-hpc-job-schedulers">
<h2>Setting up for HPC job schedulers<a class="headerlink" href="#setting-up-for-hpc-job-schedulers" title="Permalink to this headline"></a></h2>
<p>YAVSAP comes preconfigured for local use only. Yes, that’s about as ridiculous
as it sounds. What’s even more ridiculous is trying to make a configuration that
can easily be adapted to multiple HPCs and job-scheduler frameworks. There is a
compromise, however.</p>
<section id="process-labels">
<h3>Process Labels<a class="headerlink" href="#process-labels" title="Permalink to this headline"></a></h3>
<p>Rather than provide hard-coded configurations that will certainly
break, there are several Nextflow ‘labels,’ that can be used for assigning
processes to specific node queues or partitions. The labels are</p>
<dl class="simple myst">
<dt>process_low</dt><dd><p>For processes with low resource usage that take a short time</p>
</dd>
<dt>process_medium</dt><dd><p>For processes with moderate resource usage that take a moderate amount of time</p>
</dd>
<dt>process_high</dt><dd><p>For processes with high resource usage that take a moderately high amount of
time</p>
</dd>
<dt>process_long</dt><dd><p>For processes that take a long time</p>
</dd>
<dt>process_high_memory</dt><dd><p>For processes that use a lot of memory</p>
</dd>
<dt>run_local</dt><dd><p>For processes that have to be run on the login node. This label was created
specially for processes that download resources from the internet on HPCs
where compute nodes do not have internet access</p>
</dd>
</dl>
<p>Using a custom <code class="docutils literal notranslate"><span class="pre">nextflow.config</span></code> and these process labels, you can construct a
setup for your own HPC needs.</p>
</section>
<section id="example">
<h3>Example<a class="headerlink" href="#example" title="Permalink to this headline"></a></h3>
<p>As an example, here is a guide on how to set up a configuration for the
<a class="reference external" href="https://scinet.usda.gov/guide/ceres/">USDA’s SCINet Ceres cluster</a>, using the
publically available info on their website.</p>
<p>First, we see that Ceres uses SLURM and Singularity. Excellent.
Let’s set Nextflow up to use SLURM:</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">executor</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;slurm&#39;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>Some SLURM systems require an account to be passed with every job submission.
Let’s add ours just to be safe.</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">executor</span><span class="w">       </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;slurm&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">clusterOptions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;--account=millironx&#39;</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>For this example, I don’t think we’ll need to do anything special with the low,
medium, and high processes, but let’s make sure that the long and high memory
processes get submitted to partitions that can handle them.</p>
<div class="highlight-groovy notranslate"><div class="highlight"><pre><span></span><span class="n">process</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">    </span><span class="n">executor</span><span class="w">       </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;slurm&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">clusterOptions</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;--account=millironx&#39;</span><span class="w"></span>
<span class="w">    </span><span class="n">module</span><span class="w">         </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;singularity&#39;</span><span class="w"></span>
<span class="w">    </span><span class="nl">withLabel:</span><span class="w"> </span><span class="n">process_long</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">        </span><span class="n">queue</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;long&#39;</span><span class="w"></span>
<span class="w">    </span><span class="o">}</span><span class="w"></span>
<span class="w">    </span><span class="nl">withLabel:</span><span class="w"> </span><span class="n">process_high_memory</span><span class="w"> </span><span class="o">{</span><span class="w"></span>
<span class="w">        </span><span class="n">queue</span><span class="w">      </span><span class="o">=</span><span class="w"> </span><span class="s1">&#39;mem&#39;</span><span class="w"></span>
<span class="w">            </span><span class="o">}</span><span class="w"></span>
<span class="o">}</span><span class="w"></span>
</pre></div>
</div>
<p>Now, we can place this file in <code class="docutils literal notranslate"><span class="pre">$HOME/.nextflow/nextflow.config</span></code>, and these
settings will be applied every time we run the pipeline.</p>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="install.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="parameters.html" class="btn btn-neutral float-right" title="Complete Parameter Reference" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Thomas A. Christensen II.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  
<div class="rst-versions" data-toggle="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    <span class="fa fa-book"> Other Versions</span>
    v: develop
    <span class="fa fa-caret-down"></span>
  </span>
  <div class="rst-other-versions">
    <dl>
      <dt>Tags</dt>
      <dd><a href="../v0.1.0-alpha/usage.html">v0.1.0-alpha</a></dd>
      <dd><a href="../v0.1.1/usage.html">v0.1.1</a></dd>
      <dd><a href="../v0.2.0-alpha/usage.html">v0.2.0-alpha</a></dd>
      <dd><a href="../v0.2.1-alpha/usage.html">v0.2.1-alpha</a></dd>
      <dd><a href="../v0.3.0-alpha/usage.html">v0.3.0-alpha</a></dd>
      <dd><a href="../v0.3.1-alpha/usage.html">v0.3.1-alpha</a></dd>
      <dd><a href="../v0.3.2-alpha/usage.html">v0.3.2-alpha</a></dd>
      <dd><a href="../v0.3.3-alpha/usage.html">v0.3.3-alpha</a></dd>
      <dd><a href="../v0.4.0-alpha/usage.html">v0.4.0-alpha</a></dd>
      <dd><a href="../v0.5.0-alpha/usage.html">v0.5.0-alpha</a></dd>
      <dd><a href="../v0.6.0-alpha/usage.html">v0.6.0-alpha</a></dd>
      <dd><a href="../v0.6.1-alpha/usage.html">v0.6.1-alpha</a></dd>
      <dd><a href="../v0.6.2-alpha/usage.html">v0.6.2-alpha</a></dd>
      <dd><a href="../v0.6.3-alpha/usage.html">v0.6.3-alpha</a></dd>
      <dd><a href="../v0.6.4-alpha/usage.html">v0.6.4-alpha</a></dd>
      <dd><a href="../v0.6.5-alpha/usage.html">v0.6.5-alpha</a></dd>
      <dd><a href="../v0.7.0/index.html">v0.7.0</a></dd>
      <dd><a href="../v0.7.1/index.html">v0.7.1</a></dd>
      <dd><a href="../v0.8.0/index.html">v0.8.0</a></dd>
    </dl>
    <dl>
      <dt>Branches</dt>
      <dd><a href="usage.html">develop</a></dd>
    </dl>
  </div>
</div><script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>