# Output

You ran the pipeline, but now what? Here is the description of the pipeline
outputs and where to find them, but first...

## The Visualizer

Our favorite output is the (mostly) self-contained report file. The report is
contained in `report/index.html`. For the best experience, you will need all of
the files in the `report` directory, and will need to serve it with a web server
application. More details and step by step instructions are provided on the page
dedicated to {ref}`The YAVSAP Report`.

## File Output

You don't like all the fluff, huh? Here's the lowdown on the file structure you
can expect from each pipeline run.

### Report stuff

The report itself, and the alignment data needed to view it at full potential
are contained in the aptly-named `report` folder.

```text
results
└── 📁 reference
    ├── 📝 pig-serum.bam
    ├── 📝 pig-serum.bam.bai
    ├── 📝 pig-feces.bam
    ├── 📝 pig-feces.bam.bai
    ├── 📝 index.html
    ├── 📝 reference.fasta
    └── 📝 reference.fasta.fai
```

### Reference Genome

The pulled reference genome in fasta format can be found in the `reference`
folder, as well as its samtools index, under the names `result.fasta` and
`result.fasta.fai`, respectively.

```text
results
└── 📁 reference
    ├── 📝 results.fasta
    └── 📝 results.fasta.fai
```

### Classification Reports

The read classification report generated by Kraken2 that is used to filter host
reads can be found for each sample in the `classification` folder under the
filename `<samplename>.kreport`.

```text
results
└── 📁 classification
    ├── 📝 pig-serum.kreport
    └── 📝 pig-feces.kreport
```

### Alignments

Alignments for each sample against the reference genome passed to {ref}`--genome <Reference Genome Options>` can be found in the `alignment` folder in the BAM
format, as well as their sidecar index files. Alignments for each sample against
its closest match in the {ref}`--genome_list <Reference Genome Options>` can be
found in the `realignment` folder. Files are always in the `<samplename>.bam`
and `<samplename>.bam.bai` naming scheme.

```text
results
├── 📁 alignment
|   ├── 📝 pig-serum.bam
|   ├── 📝 pig-serum.bam.bai
|   ├── 📝 pig-feces.bam
|   └── 📝 pig-feces.bam.bai
└── 📁 realignment
    ├── 📝 pig-serum.bam
    ├── 📝 pig-serum.bam.bai
    ├── 📝 pig-feces.bam
    └── 📝 pig-feces.bam.bai
```

### Consensus Sequences

Consensus sequences for each sample are output to the `consensus` folder. The
filename is always `<samplename>.fasta`.

```text
results
└── 📁 consensus
    ├── 📝 pig-serum.fasta
    └── 📝 pig-feces.fasta
```

### Variant Calls

Variant calls for each sample are output to the `variants` folder. The
filename is always `<samplename>.vcf`.

```text
results
└── 📁 variants
    ├── 📝 pig-serum.vcf
    └── 📝 pig-feces.vcf
```

### Haplotypes

Haplotypes for each sample are output to the `haplotypes` folder. There is a a
fasta file containing the mutated sequences, and also a data file in YAML format
describing the haplotypes. The filename is always
`<samplename>.<ext>`.

```text
results
└── 📁 haplotypes
    ├── 📝 pig-serum.fasta
    ├── 📝 pig-serum.yaml
    ├── 📝 pig-feces.fasta
    └── 📝 pig-feces.yaml
```

### Multi-alignment and Phylogenetic Trees

The alignments of the comparison genomes, consensus sequences, and haplotypes in
fasta format are contained in the `alignment.fas` file in the `phylogenetics`
folder. Phylogenetic trees of theses comparison genomes, consensus sequences,
and haplotypes in Newick format are contained in the same folder in the file
`tree.nwk`.

```text
results
└── 📁 phylogenetics
    ├── 📝 alignment.fas
    └── 📝 tree.nwk
```

### Diagnostics

Pipeline runtime reports can be found in the `pipeline_info` folder by default, but
this can be modified using the {ref}`--tracedir <Input/Output Options>`
parameter. Four types of Nextflow traces are output.

- {ref}`Execution reports <execution-report>`
- {ref}`Execution timelines <timeline-report>`
- {ref}`Process traces <trace-report>`
- {ref}`Process graphs <dag-visualisation>`

These reports are specific to each time the pipeline is launched, and have the
date and time of execution appended, e.g.
`execution_trace_2021-10-01_12-00-00.html`.

```text
results
└── 📁 pipeline_info
    ├── 📝 execution_report_2021-09-01_12-00-00.html
    ├── 📝 execution_report_2021-10-01_12-00-00.html
    ├── 📝 execution_timeline_2021-09-01_12-00-00.html
    ├── 📝 execution_timeline_2021-10-01_12-00-00.html
    ├── 📝 execution_trace_2021-09-01_12-00-00.txt
    ├── 📝 execution_trace_2021-10-01_12-00-00.txt
    ├── 📝 pipeline_dag_2021-09-01_12-00-00.svg
    └── 📝 pipeline_dag_2021-09-01_12-00-00.svg
```
